sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -81,44 +81,36 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
     }
   }
 
+  /** A cache for the estimated statistics, such that it will only be computed once. */
+  private val statsCache = new ThreadLocal[Option[Statistics]] {
```

Comment:
Is it possible a LogicalPlan is shared by different threads? May I ask what's your concern about multi-thread in your [comment](https://github.com/apache/spark/pull/16401#issuecomment-270841604)?